docs/components/table-view.md

```diff
@@ -352,6 +366,30 @@ const getRowLink = (row: Record<string, any>): RowLink => ({
 </script>
 ```
 
+### rowKey
+
+Certain features of KTableView, such as bulk actions, require a unique identifier for each row of table data. The `rowKey` prop provides a way to define which property of the row object should serve as this unique identifier or to generate a custom unique value for each row. This prop accepts either a string or a function: `string | ((row: Record<string, any>) => string)`.
```

Comment:
```suggestion
Certain features of KTableView, such as bulk actions, require a unique identifier to utilize for each row of table data. 

The `rowKey` prop provides a way to define which property of the `row` object should serve as this unique identifier, or to generate a custom unique identifier for each row. 

The `rowKey` prop accepts either a unique string or a function that returns a unique string: `string | ((row: Record<string, any>) => string)`.
```

We should also note that the rowKey needs to be unique for the entire data set, e.g. not just the first page of data.

Reply:
I’ve been thinking that, if the row data do not already includes a unique key field, how can we generate a stable unique id from the object for the entire data set? (The case for passing a function)

Reply:
If you had a data set you wanted to take an action on (eg list of users) each user would inherently have to have a unique combination of some form of data (eg `firstName + lastName`) otherwise; how would your host app know which “thing” to act on in the first place? In this case, you could concatenate to generate a unique ID.

Alternatively, if your table data consisted of arbitrary, non-unique row data:

```js
[{ name: ‘foo’ }, { name: ‘foo’ }]
```

The host app would be required to do some sort of key generation itself, outside of the data object, to _make_ the rows unique.

Reply:
> I’ve been thinking that, if the row data do not already includes a unique key field, how can we generate a stable unique id from the object for the entire data set? (The case for passing a function)

It's a rare case but maybe we need to combine two or more fields of the row data to identify an item? e.g. `birthDate` and `name`, because people may have the same name but when combined with their date of birth, one can be uniquely determined.

Another case is that the unique key is nested, e.g. `row.xxx.id`, where we can utilize the function like this: `(row) => row.xxx.id`.

Reply:
Calculating an object’s hash (whether using its full name or other more complex cases) on the frontend cannot guarantee uniqueness, as we don’t have access to the source of truth, which typically resides in the database. Composite keys and nested keys are valid use cases, I agree. Other generated ids without predefined keys should be live with the row data itself and will only work if all the necessary data is available on the frontend (such as when data is created entirely on the frontend).

Reply:
> cannot guarantee uniqueness

correct, but this burden falls on the host application if it wants to take advantage of row actions; the same as the cache key uniqueness for swrv (in KTable)

Reply:
If the host app is going to use row actions there’s a great chance it may send request to backend and this requires the frontend to have the real unique key so I think it’s kind of inevitable. Generated ids are only useful for preserving pure frontend states, like expanded state, transition, etc. Maybe we should document that t host apps should always provide a unique row key in row data and fallback to a generated key if we’ve got nothing.

Reply:
> Maybe we should document that t host apps should always provide a unique row key in row data

This is what I actually suggested on the original PR 😅 ; essentially that each row is **required** to provide a unique property per row.

I could see where this would be overkill though for tables whereby the host app doesn't need to manipulate the data. e.g. if I just want to show a table of things, I probably don't want to have to mutate my data just to use the table component.

| Name | Count |
| :--- | :--- |
| Foo | 5 |
| Bar | 5 |

Reply:
In tables that don't care about bulk actions we don't require a unique row identifier -- it's only required for bulk actions feature.

Reply:
As for use-cases for function prop, apart from what Adam mentioned I can also envision a scenario where an `id` field in database is of type number: `[{ id: 0, name: 'foo' }, { id: 1, name: 'bar' }]` (not uncommon for SQL databases). In such scenario one might want to transform `id` a little: `:row-key="({ id }: Record<string, any>) => 'row-' + id"`.

Reply:
> It would be needed for sorting as well and any other client action that manipulates row order

Why? Is this a requirement for this PR?

Reply:
> > It would be needed for sorting as well and any other client action that manipulates row order
> 
> Why? Is this a requirement for this PR?

https://vuejs.org/guide/essentials/list.html#maintaining-state-with-key

Reply:
In fact Vue is requiring the unique keys, if we try to hide that would leave footguns where someday we are adding row states. The row states can be beyond bulk actions. Like the original issue I demonstrated in our Slack discussion, row states can be anything do not coming from props, eg. transitions, local states of components inside each row, or even text selection or scroll positions.

Reply:
For row states that don't need to persist (e.g. tracking rows that are removed from the DOM as with bulk actions) I believe we _could_ actually rely on generated unique IDs, as long as those IDs do not change while the corresponding row is visible in the DOM.

Example, if we generate a unique ID per row when the data is loaded before the virtual DOM renders the HTML, and the ID doesn't change as long as that row is in the DOM, I think Vue tracking the state of the row via the `key` would function properly so long as there aren't duplicate IDs, and the ID of any given row does not change while it is present in the DOM

Reply:
Oh, I thought the solution I pushed in [this commit](https://github.com/Kong/kongponents/pull/2427/commits/c72d08927e2830c8b7b3144a877f8ff5ad82812f) would be sufficient to not have to require `rowKey` for sorting. Do you foresee any issues with this approach?

```
:key="`table-${tableId}-row-${getRowKey(row) ?? useUniqueId()}`"
```

Reply:
If we are certain that the data is always static, not requiring unique ids is probably fine. However, in my experience, when I’m in a situation that a unique id is needed for row data, it is usually already provided by the backend API.

Reply:
If I understand correctly, `useUniqueId()` makes the key vary on each render so row state is always lost.

Reply:
> If we are certain that the data is always static

Well I don't think we always are.. for example in KTableView, since it doesn't have fetcher logic, pagination logic is handled in the host app completely - which means when user navigates through pages the host app fetches the data and just swaps `props.data` under the hood. If `loading` state is not set during the re-fetch, the `<table>` element and all it's nested children doesn't even re-render. Should we at least provide a unique key to the table element to ensure it get's re-rendered each time props.data changes?

```
<table
  :key="`tableId-${tableKey}`"
  ...
  
const tableKey = ref<number>(0)

watch(() => props.data, () => {
 tableKey.value++
 })
```

Reply:
> If I understand correctly, `useUniqueId()` makes the key vary on each render so row state is always lost.

Yes, I was thinking the keys need to be generated before the rows are rendered in the DOM, tied to the data (but not in the data in all cases)

Reply:
And it only works for either 1. static tables 2. where frontend is the source of truth.

Reply:
> Should we at least provide a unique key to the table element to ensure it get's re-rendered each time props.data changes?

I’m sorry I missed this earlier. I think it’s ok to have generated keys as a last resort but we should also surface the issue to users so they can understand why providing keys is crucial in many cases, even when they might not realize it.

Reply:
I’m also not sure you’d want the table to rerender when the data changes every time